web3/utils/datastructures.py

```diff
@@ -87,3 +89,81 @@ def __eq__(self, other):
             return self.__dict__ == dict(other)
         else:
             return False
+
+
+class NamedElementStack:
```

Comment:
This feels like it *should* inherit from at least one of the `collections.abc` classes.  Maybe any or all of:

* `Iterable`
* `Sequence`
* `Mapping`

Reply:
Oh, and for python2, this needs to inherit from `object` at a minimum.

Reply:
Hm, I started this is a subclass of an `OrderedDict`, and now I forget exactly why that was a bad idea. I think it was that adding via `stack[key] = val` seemed like an unnatural way to say that val is now at the top of the stack.

I'm on board with `Sequence`, which itself inherits from `Iterable`.

Reply:
Actually, even `OrderedDict` isn't a sequence, so I'll stick with `Iterable` for now. If we have a good reason to add `Mapping` later, we can add it.

Reply:
Actually, non-mutable `Mapping` looks good.